File: fons/norma-py/yaml.py
"""yaml.py - YAML Encoding/Decoding Implementation

Native Python implementation of the YAML interface.
Uses PyYAML.

Verb meanings:
  - pange (compose): serialize value to YAML string
  - necto (bind): bind multiple documents into multi-doc YAML
  - solve (untangle): parse YAML string to value
  - tempta (try): attempt to parse, return None on error
  - collige (gather): gather all documents from multi-doc YAML
"""

from typing import Any, Optional

import yaml as _yaml


# =============================================================================
# SERIALIZATION
# =============================================================================

def pange(valor: Any) -> str:
    """Serialize value to YAML string."""
    return _yaml.dump(valor, default_flow_style=False, allow_unicode=True)


def necto(documenta: list[Any]) -> str:
    """Bind multiple documents into multi-doc YAML string."""
    return _yaml.dump_all(documenta, default_flow_style=False, allow_unicode=True)


# =============================================================================
# PARSING
# =============================================================================

def solve(yaml_str: str) -> Any:
    """Parse YAML string to value (raises on error)."""
    return _yaml.safe_load(yaml_str)


def tempta(yaml_str: str) -> Optional[Any]:
    """Attempt to parse YAML string (returns None on error)."""
    try:
        return _yaml.safe_load(yaml_str)
    except _yaml.YAMLError:
        return None


def collige(yaml_str: str) -> list[Any]:
    """Gather all documents from multi-doc YAML string."""
    return list(_yaml.safe_load_all(yaml_str))


# =============================================================================
# TYPE CHECKING
# =============================================================================

def est_nihil(valor: Any) -> bool:
    """Check if value is null."""
    return valor is None


def est_bivalens(valor: Any) -> bool:
    """Check if value is boolean."""
    return isinstance(valor, bool)


def est_numerus(valor: Any) -> bool:
    """Check if value is number."""
    return isinstance(valor, (int, float)) and not isinstance(valor, bool)


def est_textus(valor: Any) -> bool:
    """Check if value is string."""
    return isinstance(valor, str)


def est_lista(valor: Any) -> bool:
    """Check if value is array/sequence."""
    return isinstance(valor, list)


def est_tabula(valor: Any) -> bool:
    """Check if value is object/mapping."""
    return isinstance(valor, dict)


# =============================================================================
# VALUE EXTRACTION
# =============================================================================

def ut_textus(valor: Any, def_val: str) -> str:
    """Extract as string with default."""
    return valor if isinstance(valor, str) else def_val


def ut_numerus(valor: Any, def_val: float) -> float:
    """Extract as number with default."""
    if isinstance(valor, (int, float)) and not isinstance(valor, bool):
        return float(valor)
    return def_val


def ut_bivalens(valor: Any, def_val: bool) -> bool:
    """Extract as boolean with default."""
    return valor if isinstance(valor, bool) else def_val


# =============================================================================
# VALUE ACCESS
# =============================================================================

def cape(valor: Any, clavis: str) -> Any:
    """Get value by key (returns None if missing)."""
    if isinstance(valor, dict):
        return valor.get(clavis)
    return None


def carpe(valor: Any, index: int) -> Any:
    """Pluck value by array index (returns None if out of bounds)."""
    if isinstance(valor, list) and 0 <= index < len(valor):
        return valor[index]
    return None


def inveni(valor: Any, via: str) -> Any:
    """Find value by dotted path (returns None if not found)."""
    parts = via.split(".")
    current = valor

    for part in parts:
        if current is None:
            return None
        if isinstance(current, dict):
            current = current.get(part)
        elif isinstance(current, list):
            try:
                idx = int(part)
                if 0 <= idx < len(current):
                    current = current[idx]
                else:
                    return None
            except ValueError:
                return None
        else:
            return None

    return current

File: radix/crates/norma/yaml.rs
//! yaml.rs - YAML Encoding/Decoding Implementation
//!
//! Native Rust implementation of the YAML interface.
//! Uses serde_yaml.
//!
//! Verb meanings:
//!   - pange (compose): serialize value to YAML string
//!   - necto (bind): bind multiple documents into multi-doc YAML
//!   - solve (untangle): parse YAML string to value
//!   - tempta (try): attempt to parse, return None on error
//!   - collige (gather): gather all documents from multi-doc YAML

use serde::Deserialize;
use serde_yaml::Value;

// =============================================================================
// SERIALIZATION
// =============================================================================

/// Serialize value to YAML string
pub fn pange(valor: &Value) -> String {
    serde_yaml::to_string(valor).expect("failed to serialize YAML")
}

/// Bind multiple documents into multi-doc YAML string
pub fn necto(documenta: &[Value]) -> String {
    documenta
        .iter()
        .map(|doc| serde_yaml::to_string(doc).expect("failed to serialize YAML"))
        .collect::<Vec<_>>()
        .join("---\n")
}

// =============================================================================
// PARSING
// =============================================================================

/// Parse YAML string to value (panics on error)
pub fn solve(yaml: &str) -> Value {
    serde_yaml::from_str(yaml).expect("failed to parse YAML")
}

/// Attempt to parse YAML string (returns None on error)
pub fn tempta(yaml: &str) -> Option<Value> {
    serde_yaml::from_str(yaml).ok()
}

/// Gather all documents from multi-doc YAML string
pub fn collige(yaml: &str) -> Vec<Value> {
    let mut docs = Vec::new();
    for doc in serde_yaml::Deserializer::from_str(yaml) {
        if let Ok(value) = Value::deserialize(doc) {
            docs.push(value);
        }
    }
    docs
}

// =============================================================================
// TYPE CHECKING
// =============================================================================

/// Check if value is null
pub fn est_nihil(valor: &Value) -> bool {
    valor.is_null()
}

/// Check if value is boolean
pub fn est_bivalens(valor: &Value) -> bool {
    valor.is_bool()
}

/// Check if value is number
pub fn est_numerus(valor: &Value) -> bool {
    valor.is_number()
}

/// Check if value is string
pub fn est_textus(valor: &Value) -> bool {
    valor.is_string()
}

/// Check if value is array/sequence
pub fn est_lista(valor: &Value) -> bool {
    valor.is_sequence()
}

/// Check if value is object/mapping
pub fn est_tabula(valor: &Value) -> bool {
    valor.is_mapping()
}

// =============================================================================
// VALUE EXTRACTION
// =============================================================================

/// Extract as string with default
pub fn ut_textus(valor: &Value, def_val: &str) -> String {
    valor
        .as_str()
        .map(|s| s.to_string())
        .unwrap_or_else(|| def_val.to_string())
}

/// Extract as number with default
pub fn ut_numerus(valor: &Value, def_val: f64) -> f64 {
    valor.as_f64().unwrap_or(def_val)
}

/// Extract as boolean with default
pub fn ut_bivalens(valor: &Value, def_val: bool) -> bool {
    valor.as_bool().unwrap_or(def_val)
}

// =============================================================================
// VALUE ACCESS
// =============================================================================

/// Get value by key (returns null if missing)
pub fn cape(valor: &Value, clavis: &str) -> Value {
    valor
        .get(clavis)
        .cloned()
        .unwrap_or(Value::Null)
}

/// Pluck value by array index (returns null if out of bounds)
pub fn carpe(valor: &Value, index: usize) -> Value {
    valor
        .get(index)
        .cloned()
        .unwrap_or(Value::Null)
}

/// Find value by dotted path (returns null if not found)
pub fn inveni(valor: &Value, via: &str) -> Value {
    let parts: Vec<&str> = via.split('.').collect();
    let mut current = valor;

    for part in parts {
        match current {
            Value::Mapping(map) => {
                let key = Value::String(part.to_string());
                current = match map.get(&key) {
                    Some(v) => v,
                    None => return Value::Null,
                };
            }
            Value::Sequence(arr) => {
                let idx: usize = match part.parse() {
                    Ok(i) => i,
                    Err(_) => return Value::Null,
                };
                current = match arr.get(idx) {
                    Some(v) => v,
                    None => return Value::Null,
                };
            }
            _ => return Value::Null,
        }
    }

    current.clone()
}
